services/src/diagnostics.cpp

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2017 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "diagnostics.h"
+
+#include "spark_wiring_vector.h"
+
+#include "system_error.h"
+
+#include <algorithm>
+
+namespace {
+
+using namespace spark;
+
+class Diagnostics {
+public:
+    Diagnostics() :
+            enabled_(0) { // The service is disabled initially
```

Comment:
what's the rationale for having enable/disable for the service? If it's to save expensive computation then perhaps we should move the enable flag to the diag_source?

Reply:
The rationale is to avoid a race condition, which is possible when the system receives a request or a call querying diagnostic data but data sources are still being registered. The system explicitly enables the diagnostic service after invoking all global constructors in all system parts [here](https://github.com/spark/firmware/blob/feature/diagnostics/system/src/main.cpp#L657)

Reply:
Thanks, that makes sense. The name `enabled` is slightly counter-intuitive and that you cannot add sources once it is enabled. On first reading, I was sure I understood it as the opposite - the error is raised when adding sources and the service is not enabled. 

Perhaps we should call it running, or started?  Alternatively, you could add a comment to explain, since normally things are not usable while disabled, while here it's the opposite.

Reply:
👍